### PR TITLE
KV Store: Add repository state

### LINKED ISSUE #3813: KV Graveler: Add repository unique identifier and states
First part of https://github.com/treeverse/lakeFS/issues/3713

### PR DESCRIPTION
**Pull request by N-o-Z**
Added state, uid and modified create flow. Delete flow will be updated in https://github.com/treeverse/lakeFS/issues/3713
closes #3813

### CODE REVIEW

File: pkg/graveler/ref/manager.go [hunk 1]
```diff
@@ -84,15 +103,19 @@ func (m *KVManager) getRepositoryRec(ctx context.Context, repositoryID graveler.
...
 		return nil, err
 	}
-	return graveler.RepoFromProto(&repo).Repository, nil
+
+	switch repo.State {
+	case graveler.RepositoryStateActive:
+		return repo, nil
+	case graveler.RepositoryStateDeleted:
+		return nil, graveler.ErrRepositoryInDeletion
+	default:
```
Comment (itaiad200) [inline pkg/graveler/ref/manager.go:116]: We have 1 more state. What about it?

File: pkg/graveler/ref/manager.go [hunk 2]
```diff
@@ -140,10 +174,7 @@ func (m *KVManager) ListRepositories(ctx context.Context) (graveler.RepositoryIt
...
 	// TODO: delete me
 	// temp code to align with DB manager. Delete once https://github.com/treeverse/lakeFS/issues/3640 is done
-	_, err := m.GetRepository(ctx, repositoryID)
-	if errors.Is(err, kv.ErrNotFound) {
-		return graveler.ErrRepositoryNotFound
-	}
+	_, err := m.getRepository(ctx, repositoryID)
 	if err != nil {
 		return err
 	}
```
Comment (itaiad200) [inline pkg/graveler/ref/manager.go:180]: what is the below TODO?

File: pkg/graveler/ref/manager.go [hunk 3]
```diff
@@ -111,7 +135,7 @@ func (m *KVManager) createBareRepository(ctx context.Context, repositoryID grave
 	return nil
 }
 
-func (m *KVManager) CreateRepository(ctx context.Context, repositoryID graveler.RepositoryID, repository graveler.Repository, token graveler.StagingToken) error {
+func (m *KVManager) CreateRepository(ctx context.Context, repositoryID graveler.RepositoryID, repository graveler.Repository) error {
```
Comment (itaiad200) [inline pkg/graveler/ref/manager.go:138]: We said we'll create the repo last...

File: pkg/graveler/ref/manager.go [hunk 4]
```diff
@@ -71,9 +72,27 @@ func NewKVRefManager(executor batch.Batcher, kvStore kv.StoreMessage, addressPro
 	}
 }
 
+func (m *KVManager) updateRepoState(ctx context.Context, repo *graveler.RepositoryRecord, state graveler.RepositoryState) error {
```
Comment (itaiad200) [inline pkg/graveler/ref/manager.go:75]: I think some of these usages should be `SetIf`

**Review by itaiad200 — CHANGES_REQUESTED** [review]
Thanks! Simple and clean implementation.
I do think we need to add tests since the repo creation and deletion are becoming more complicated & fragile.

File: pkg/graveler/ref/manager.go (hunk 1 above)
Comment (N-o-Z) [reply pkg/graveler/ref/manager.go:116]: Currently we return the default error for state other than in-deletion and active.
This can happen for example when creating repository the second time while it is already being created, the chances of that are very small, thus I prefer not creating a new error type for it

File: pkg/graveler/ref/manager.go (hunk 2 above)
Comment (N-o-Z) [reply pkg/graveler/ref/manager.go:180]: Delete flow will be addressed in https://github.com/treeverse/lakeFS/issues/3713

File: pkg/graveler/ref/manager.go (hunk 4 above)
Comment (N-o-Z) [reply pkg/graveler/ref/manager.go:75]: For now, I don't see a scenario for SetIf, it will also require us to have another Get or pass the predicate to this function.
I suggest handling this when we have a valid scenario for SetIf

**Comment by N-o-Z** [comment]
Will add test for state once we complete implementation of repository delete

File: pkg/graveler/ref/kv_migrate_test.go [hunk 5]
```diff
@@ -168,6 +167,9 @@ func verifyMigrationResults(t *testing.T, ctx context.Context, kvMgr, dbMgr grav
 		dbr, err := dbMgr.GetRepository(ctx, repo.RepositoryID)
 		require.NoError(t, err)
 		dbr.CreationDate = dbr.CreationDate.UTC()
+		// Fill DB missing fields
+		dbr.State = graveler.RepositoryState_ACTIVE
+		dbr.InstanceUID = kvr.InstanceUID
```
Comment (itaiad200) [inline pkg/graveler/ref/kv_migrate_test.go:172]: Can we check it's not empty?

File: pkg/graveler/graveler.go [hunk 6]
```diff
@@ -191,6 +192,18 @@ type Repository struct {
 	StorageNamespace StorageNamespace `db:"storage_namespace"`
 	CreationDate     time.Time        `db:"creation_date"`
 	DefaultBranchID  BranchID         `db:"default_branch"`
+	State            RepositoryState
```
Comment (itaiad200) [inline pkg/graveler/graveler.go:195]: I think this is the first occurrence that we use the protobuf type in the golang struct. Why did you remove the golang type?

File: pkg/graveler/graveler.go [hunk 7]
```diff
@@ -842,26 +855,27 @@ func (g *KVGraveler) GetRepository(ctx context.Context, repositoryID RepositoryI
...
 
 func (g *KVGraveler) CreateRepository(ctx context.Context, repositoryID RepositoryID, storageNamespace StorageNamespace, branchID BranchID) (*Repository, error) {
-	repo := Repository{
-		StorageNamespace: storageNamespace,
-		CreationDate:     time.Now(),
-		DefaultBranchID:  branchID,
+	_, err := g.RefManager.GetRepository(ctx, repositoryID)
+	if err != nil && !errors.Is(err, ErrRepositoryNotFound) {
+		return nil, err
 	}
```
Comment (itaiad200) [inline pkg/graveler/graveler.go:861]: What's the purpose of this test if it can happen anyhow during creation? I think we should expect `kv.ErrPredicateFailed` from the creation and act on that.

File: pkg/graveler/graveler.go [hunk 8]
```diff
@@ -842,26 +855,27 @@ func (g *KVGraveler) GetRepository(ctx context.Context, repositoryID RepositoryI
...
 
 func (g *KVGraveler) CreateBareRepository(ctx context.Context, repositoryID RepositoryID, storageNamespace StorageNamespace, defaultBranchID BranchID) (*Repository, error) {
-	repo := Repository{
-		StorageNamespace: storageNamespace,
-		CreationDate:     time.Now(),
-		DefaultBranchID:  defaultBranchID,
+	_, err := g.RefManager.GetRepository(ctx, repositoryID)
+	if err != nil && !errors.Is(err, ErrRepositoryNotFound) {
+		return nil, err
 	}
```
Comment (itaiad200) [inline pkg/graveler/graveler.go:875]: Same here.

File: pkg/graveler/graveler.go [hunk 9]
```diff
@@ -191,6 +192,18 @@ type Repository struct {
...
 	CreationDate     time.Time        `db:"creation_date"`
 	DefaultBranchID  BranchID         `db:"default_branch"`
+	State            RepositoryState
+	InstanceUID      string
+}
+
+func NewRepository(storageNamespace StorageNamespace, defaultBranchID BranchID) Repository {
+	return Repository{
+		StorageNamespace: storageNamespace,
+		CreationDate:     time.Now().UTC(),
```
Comment (itaiad200) [inline pkg/graveler/graveler.go:202]: I don't think UTC matters here.

File: pkg/graveler/graveler.go [hunk 10]
```diff
@@ -191,6 +192,18 @@ type Repository struct {
 	StorageNamespace StorageNamespace `db:"storage_namespace"`
 	CreationDate     time.Time        `db:"creation_date"`
 	DefaultBranchID  BranchID         `db:"default_branch"`
+	State            RepositoryState
+	InstanceUID      string
```
Comment (itaiad200) [inline pkg/graveler/graveler.go:196]: We need documenetation for these non-straightforward properties.

**Review by itaiad200 — APPROVED** [review]
Few comments to fix but nothing blocking, thanks!

File: pkg/graveler/graveler.go (hunk 6 above)
Comment (N-o-Z) [reply pkg/graveler/graveler.go:195]: No need for redundant declaration of of enums - it will require unnecessary maintenance when adding / removing states. We only duplicate proto structs to keep backwards compatibility with DB. Optimally we would never have needed to create golang structs and perform the conversions all the time

File: pkg/graveler/graveler.go (hunk 9 above)
Comment (N-o-Z) [reply pkg/graveler/graveler.go:202]: It matters - we should not keep time in localized format since protobuf doesn't save time localization information

Comment (itaiad200) [reply pkg/graveler/graveler.go:202]: from `timestamp.pb.go`:
```
func New(t time.Time) *Timestamp {
	return &Timestamp{Seconds: int64(t.Unix()), Nanos: int32(t.Nanosecond())}
}
```

so protobuf's library got your back :)

Comment (N-o-Z) [reply pkg/graveler/graveler.go:202]: It returns a protobuf timestamp object, we are using the golang variation

Comment (itaiad200) [reply pkg/graveler/graveler.go:202]: Right - I thought you were worried about the conversation (I can't think of any other reason to use UTC here)

File: pkg/graveler/graveler.go (hunk 10 above)
Comment (N-o-Z) [reply pkg/graveler/graveler.go:196]: I think the convention we're trying to hold is always to treat time in UTC. @nopcoder correct me if I'm wrong

Comment (N-o-Z) [reply pkg/graveler/graveler.go:196]: Done

File: pkg/graveler/ref/kv_migrate_test.go (hunk 5 above)
Comment (N-o-Z) [reply pkg/graveler/ref/kv_migrate_test.go:172]: Done

File: pkg/graveler/graveler.go (hunk 7 above)
Comment (N-o-Z) [reply pkg/graveler/graveler.go:861]: We want to provide the user with informative error in case of different repository states.
For example, if repository is "in deletion". The user will know to either wait for deletion to complete and retry or call delete again
